kv/kvserver: TestLeaderAfterSplit failed

kv/kvserver.TestLeaderAfterSplit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9629661?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9629661?buildTab=artifacts#/) on release-23.1.0 @ [d9427da31576a61b9390d86258fd7ad4c35c88d5](https://github.com/cockroachdb/cockroach/commits/d9427da31576a61b9390d86258fd7ad4c35c88d5):


```
=== RUN   TestLeaderAfterSplit
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaderAfterSplit2506029295
    test_log_scope.go:79: use -show-logs to present logs inline
    client_split_test.go:2225: result is ambiguous: replica unavailable: (n1,s1):1 unable to serve request to r63:{m-/Table/0} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5]: closed timestamp: 1681646464.227352198,0 (2023-04-16 12:01:04); raft status: {"id":"1","term":5,"vote":"0","commit":10,"lead":"0","raftState":"StatePreCandidate","applied":10,"progress":{},"leadtransferee":"0"}: have been waiting 63.00s for slow proposal Increment ["z",/Min)
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaderAfterSplit2506029295
--- FAIL: TestLeaderAfterSplit (71.99s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaderAfterSplit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27055


Epic CRDB-27234

kv/kvserver.TestLeaderAfterSplit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10087289?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10087289?buildTab=artifacts#/) on release-23.1.0 @ [358e0d87912365b8976c55ab9b3292e999cf720d](https://github.com/cockroachdb/cockroach/commits/358e0d87912365b8976c55ab9b3292e999cf720d):


```
=== RUN   TestLeaderAfterSplit
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaderAfterSplit1223791309
    test_log_scope.go:79: use -show-logs to present logs inline
    client_split_test.go:2225: result is ambiguous: replica unavailable: (n1,s1):1 unable to serve request to r64:{m-/Table/0} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5]: closed timestamp: 1684151101.862621089,0 (2023-05-15 11:45:01); raft status: {"id":"1","term":5,"vote":"0","commit":10,"lead":"0","raftState":"StatePreCandidate","applied":10,"progress":{},"leadtransferee":"0"}: have been waiting 63.00s for slow proposal Increment ["z",/Min)
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeaderAfterSplit1223791309
--- FAIL: TestLeaderAfterSplit (71.90s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaderAfterSplit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -2185,24 +2185,34 @@ func TestStoreRangeSplitRaceUninitializedRHS(t *testing.T) {
 	}
 }
 
-// TestLeaderAfterSplit verifies that a raft group created by a split
-// elects a leader without waiting for an election timeout.
+// TestLeaderAfterSplit verifies that a raft group created by a split elects a
+// leader without waiting for an election timeout. It also tests that we don't
+// get an election tie, because we only allow the leaseholder to campaign if
+// there is a valid lease.
 func TestLeaderAfterSplit(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	ctx := context.Background()
-	tc := testcluster.StartTestCluster(t, 3,
-		base.TestClusterArgs{
-			ReplicationMode: base.ReplicationManual,
-			ServerArgs: base.TestServerArgs{
-				RaftConfig: base.RaftConfig{
-					RaftElectionTimeoutTicks: 1000000,
-				},
+	// Timing-sensitive test, disable under deadlock and stressrace.
+	skip.UnderDeadlock(t)
+	skip.UnderStressRace(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second) // time out early
+	defer cancel()
+
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs: base.TestServerArgs{
+			RaftConfig: base.RaftConfig{
+				RangeLeaseDuration:       time.Hour, // don't expire leases
+				RaftElectionTimeoutTicks: 1000000,   // disable elections
 			},
-		})
+		},
+	})
 	defer tc.Stopper().Stop(ctx)
+
 	store := tc.GetFirstStoreFromServer(t, 0)
+	sender := tc.Servers[0].DistSender()
 
 	leftKey := roachpb.Key("a")
 	splitKey := roachpb.Key("m")
@@ -2212,20 +2222,17 @@ func TestLeaderAfterSplit(t *testing.T) {
 	require.NotNil(t, repl)
 	tc.AddVotersOrFatal(t, repl.Desc().StartKey.AsRawKey(), tc.Targets(1, 2)...)
 
-	splitArgs := adminSplitArgs(splitKey)
-	if _, pErr := kv.SendWrapped(ctx, tc.Servers[0].DistSender(), splitArgs); pErr != nil {
-		t.Fatal(pErr)
-	}
+	// Split the range.
+	_, pErr := kv.SendWrapped(ctx, sender, adminSplitArgs(splitKey))
+	require.NoError(t, pErr.GoError())
 
-	incArgs := incrementArgs(leftKey, 1)
-	if _, pErr := kv.SendWrapped(ctx, tc.Servers[0].DistSender(), incArgs); pErr != nil {
-		t.Fatal(pErr)
-	}
+	// Make sure both the LHS and RHS can replicate a write request. This will
+	// time out if the RHS can't elect a Raft leader.
+	_, pErr = kv.SendWrapped(ctx, sender, incrementArgs(leftKey, 1))
+	require.NoError(t, pErr.GoError())
 
-	incArgs = incrementArgs(rightKey, 2)
-	if _, pErr := kv.SendWrapped(ctx, tc.Servers[0].DistSender(), incArgs); pErr != nil {
-		t.Fatal(pErr)
-	}
+	_, pErr = kv.SendWrapped(ctx, sender, incrementArgs(rightKey, 1))
+	require.NoError(t, pErr.GoError())
 }
 
 func BenchmarkStoreRangeSplit(b *testing.B) {

